README.rst

```diff
@@ -26,3 +26,371 @@ Run `tox` to run the whole test suite:
+   print(ctrl.term.before)         # print the command result
+   ctrl.stop_term()                # close the terminal
+
+Creating a RIOTCtrl object is done via environments. If empty then all
+configuration will come from the target application makefile. But any
+Make environment variable can be overridden, for example setting
+``BOARD`` to a target ``BOARD`` which is not the default for that
+application.
+
+Any make target used on RIOT devices can be used on the abstraction
+like: ``make flash`` => ``ctrl.make_run(['flash'])``.
+
+``ctrl.start_term()`` (``make term``\ ’s alter ego) by default spawns a
+pexpect child application. From there interactions with the application
+under use can be atomized. In the example below the output of the
+``"help"`` command is captured:
+
+ShellInteractions
+~~~~~~~~~~~~~~~~~
+
+RIOTCtrl provides a minimal extensions by using pexpect replwrap: “[A]
+Generic wrapper for read-eval-print-loops, a.k.a. interactive shells”.
+This implements the perfect wrapper for RIOT shell commands since it
+will wait for a command to finish before returning its output.
+
+RIOT already provides a ``ShellInteraction`` for the ``"help"`` command,
+it can be imported as ``from riotctrl_shell.sys import Help``.
+
+The previous example can be re-written using ``ShellInteraction``:
+
+.. code:: python
+
+   from riotctrl.ctrl import RIOTCtrl
+   from riotctrl.shell import ShellInteraction
+
+   env = {'BOARD': 'native'}
+   # if not running from the application directory the a path must be provided
+   ctrl = RIOTCtrl(env=env, application_directory='.')
+   # flash the application
+   ctrl.flash()                     # alias for ctrl.make_run(['flash'])
+   # shell interaction instance
+   shell = ShellInteraction(ctrl)
+   shell.start_term()               # start a serial terminal
+   print(shell.cmd("help"))         # print the command result
+   shell.stop_term()                # close the terminal
+
+or using the already provided ``Help`` ``ShellInteraction``:
+
+.. code:: python
+
+   class Help(ShellInteraction):
+       """Help ShellInteraction"""
+       @ShellInteraction.check_term
+       def help(self, timeout=-1, async_=False):
+           """Sends the reboot command via the terminal"""
+           return self.cmd("help", timeout, async_)
+
+.. code:: python
+
+   from riotctrl.ctrl import RIOTCtrl
+   from riotctrl_shell.sys import Help
+
+   env = {'BOARD': 'native'}
+   # if not running from the application directory the a path must be provided
+   ctrl = RIOTCtrl(env=env, application_directory='.')
+   # flash the application
+   ctrl.flash()                     # alias for ctrl.make_run(['flash'])
+   # shell interaction instance, Help uses the @ShellInteraction.check_term
+   # decorator, it will start the terminal if its not yet running, and close
+   # it after the command ends
+   shell = Help(ctrl)              #
+   print(shell.help())             # print the command result
+
+Writing SAUL ShellInteraction
```

Comment:
This section and the next could out-date pretty fast, as soon as there is a SAUL shellinteraction in upstream... Maybe do this on the basis of a shell command defined in this text. For instance, "Given a main file 'main.c' \<print main.c with some example shell command>, we can write the following ShellInteraction extension.

Reply:
Does its really matter? I'm giving the commands here, with the expected output, even if output changes in the future the example would remain valid.

Reply:
Would just be more future-proof IMHO, to document this independent from RIOT-OS/RIOT. So yes, it matters.

Reply:
Done